### PR TITLE
Messenger System

### PR DESCRIPTION
**Pull request by johmarjac**
Allright,

dont merge this yet, I wanna discuss that system here first but showing off what I have so far.

### CODE REVIEW

File: src/Rhisis.World/Systems/Messenger/EventArgs/AddFriendCancelEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+﻿using Rhisis.World.Game.Core;
+
+namespace Rhisis.World.Systems.Messenger.EventArgs
+{
+    public class AddFriendCancelEventArgs : SystemEventArgs
+    {
+        public int LeaderId { get; }
+        public int MemberId { get; }
```
[the remark this answers is not in the document]
Comment (johmarjac) [reply src/Rhisis.World/Systems/Messenger/EventArgs/AddFriendCancelEventArgs.cs]: I totally agree

File: src/Rhisis.World/Systems/Messenger/MessengerSystem.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                .FirstOrDefault() as IPlayerEntity;
+
+            OnAddFriendRequest(playerEntity, member);
+        }
+
+        private void OnAddFriendRequest(IPlayerEntity playerEntity, IPlayerEntity memberEntity)
+        {
+            if (memberEntity == null)
+            {
+                Logger.Error($"memberEntity is null.");
```
[the remark this answers is not in the document]
Comment (johmarjac) [reply src/Rhisis.World/Systems/Messenger/MessengerSystem.cs]: Okay, no problem ;)

**Comment by johmarjac** [comment]
I see, I havent had that on my mind. Does the same apply to making friend requests? (Can they be made over different channels too? )

**Comment by johmarjac** [comment]
@Eastrall Yes I dont remember either xD

I will try that out when I am home. Today might be the last day I have time for that as I go on holiday tomorrow for 2 weeks. So dont wonder there wont be progress in next 2 weeks :D

**Comment by johmarjac** [comment]
@Eastrall In case, sending friend requests over different channels does not work on official files, do we plan to implement that as an additional feature ? Because I think that can work when just implementing that the correct way on the server side.

**Comment by deadphoenix8091** [comment]
Adding friends works across multiple servers using official files.
Also seeing that the client and the server knows which server the other player is currently connected to, the private message could be sent directly to the other server instead of sending it to every server which would be alot more internal traffic that the other servers have to discard anyway.

**Comment by Eastrall** [comment]
Thank you for the information @deadphoenix8091. You are right about what you are saying. Instead of broadcasting to every channel, we could keep a list of connected players on the cluster server with the id of the channel he is connected to. So, on the cluster server we will directly fetch the correct server and send the request.

**Comment by Skeatwin** [comment]
The question is how it's implemented on client side ?
If we take as hypothesis that a connected players list isstored on cluster server.

On private message event:
**1. Client handle a direct connection ?** In that case, on new player connection, server have to send to every players a refresh list of connected friends with channel ID (or IP).
**2. Client send always to parent cluster ?** Cluster server will find when other player is connected and send to World server.
**3. (Variante) Client handle a direct connection only if player is on the same channel, otherwise client send to parent cluster.**

**Comment by Eastrall** [comment]
Nothing changes on client side. Everything will be made on server side with the ISC.
For exemple, a client connects the world server. We send a message to the cluster telling that player `XXXX` is connected on WorldServer with id `Y`.

So when we make a friend request, we will first check if the player is not on the same channel, if it is, well, just execute the send friend request system and otherwise, send a message to the cluster server via ISC that will send the friend request to the right channel.

**Comment by deadphoenix8091** [comment]
When a user connects to a specific world server there needs to be some kind of ISC event to tell ALL other Servers about the connection. Which will in turn trigger an event message to the clients of friends. Each server should also keep this information in Memory as it will also be needed for the guild and party system. (Something like a user to server map which can be queried directly)

Also keep in mind that direct /say and /whisper messages also work across servers.

**Comment by Skeatwin** [comment]
Yes, I know ! We don't want to change the client, but we need to know what are the handle possibilities.

Does the client have a World Server IP cache list and does it send directly packet to player target channel ? Or send always to his own connected world server ? Maybe the client deal only with cluster server for privates messages, I don't know. But of course, if we want to implement this system we need to know clearly how the client currently handle-it.

**Comment by Eastrall** [comment]
Again, the client doesn't have anything to do here. The client is "just" a visual representation of what is happening on the server. The server sends informations to the client, and the client displays it.
In this case, everything is handled by the server side that will redirect the message to the good world server and then this world server will send the message to the appropriate client.

**Comment by Skeatwin** [comment]
No, definitely not and you said-it by your self: "the client (send/receive informations and) displays-it" 😛
 
We are agree, we don't want to change the client. SO, we have to fit perfectly with informations waiting by it, and information that it will send to us. That's why **I would like someone explain how the "messenger system" is handle currently by the client**, in order to think to the best way to handle it on server side. 
I know all business treatment is (should be) done by the server, but the client has a contract that we need to respect and follow.

For exemple, are you really sure if a chat packet is send to World server and not to Cluster server ?
What are the informations waiting and sending by the client, which packet type/flags ?
What I'm saying from begining is that we need to know exacly which are the interactions between client and server before to think to this system. It's not a simple system (like inventory, or mover), it seems to have a lot of dependancies (players, friends, guilds, World servers, broadcasting ?...). 

Hope I've been more clear this time. I know it's diffcult and I'm not an expert into explainations ^^

**Comment by Eastrall** [comment]
All game packets are sent to the WorldServer. None of them are sent to the Cluster. When I was talking about the Cluster was only for the ISC part. 😄 
I believe when the client sends a friend request, they send the destination player id and then we need to handle it has followed:
- If player is on this server **then** send the friend request to this player
- If player is **not** on this server **then** send the request to cluster which will **check** where the player is and **send** the information to the correct **WorldServer** where the player is. And this **WorldServer** is going to send the friend request to the player.

File: src/Rhisis.World/Systems/Messenger/EventArgs/AddFriendCancelEventArgs.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (CineXMike) [reply src/Rhisis.World/Systems/Messenger/EventArgs/AddFriendCancelEventArgs.cs]: Indeed!

**Comment by Eastrall** [comment]
@johmarjac You are going to have some issues while merging because the system definition changed. Check PR #123 to see how to define systems. Don't worry, it's easier than it was before. 😉

**Comment by johmarjac** [comment]
@Eastrall Yeah, thanks for the hint. Unfortunately I can´t get back to work at Rhisis anyways in the next weeks as I am having difficulties at work right now. When I am able to work again on Rhisis I will probably start from scratch, it was´t that much I did yet and I even lost overview anyways already :D

**Comment by Eastrall** [comment]
Alright @johmarjac. When you'll get back on the project, maybe the ISC communication interface will be done and you'll be able to make broadcasts through channels. 😄

**Comment by johmarjac** [comment]
lol, no its a very old PR, it needs to be refactored anyways because of the ISC to be able to send friend requests across channels.

**Comment by Eastrall** [comment]
Closing PR.

**Pull request closed without merging**